### DISCUSSION TITLE
Cannot authenticate with x509 without username and password

### DISCUSSION THREAD

**Discussion by Aleksandar1932** [rabbitmq/rabbitmq-server#8827] [Other]
### Describe the bug

I've successfully configured RabbitMQ with for TLS following the official docs.

My config looks like:
```
listeners.ssl.default = 5671

log.file.formatter = json
log.console.formatter = json

ssl_options.cacertfile = /etc/rabbitmq/ssl/ca_certificate.pem
ssl_options.certfile = /etc/rabbitmq/ssl/server_certificate.pem
ssl_options.keyfile = /etc/rabbitmq/ssl/server_key.pem
ssl_options.verify = verify_peer
ssl_options.fail_if_no_peer_cert = true
ssl_cert_login_from = common_name

mqtt.ssl_cert_login = true
mqtt.listeners.ssl.default = 8883
mqtt.allow_anonymous = true

auth_mechanisms.1 = EXTERNAL
```

When I want to authenticate with an MQTT client it's not possible without specifying a valid username and password. As per https://www.rabbitmq.com/mqtt.html it is stated `Clients must not supply username and password.` 

Is this behaviour expected?

### Reproduction steps

1. Try to login with client cert and ca
2. bad username and password

### Expected behavior

login without username/password, just with x509 certificate

**Comment by michaelklishin** [comment]
When x.509 certificates are used, the username is extracted from one of the certificate fields and the password field is ignored.

The MQTT guide likely means that those credentials will not be used as if it were the standard authN mechanism.

**Comment by michaelklishin** [comment]
What you claim to be doing is perfectly possible and hasn't changed in years. 

Perhaps you haven't restarted all nodes after updating `rabbitmq.conf`. Or your certificates do not have the appropriate x.509 extensions. See what [tls-gen](https://github.com/rabbitmq/tls-gen) generates, for example, we have been using it across all of our test suites for over a decade.

See [How to Inspect and Verify Effective Configuration of a Running Node](https://rabbitmq.com/configure.html#verify-configuration-effective-configuration) and [Troubleshooting TLS](https://rabbitmq.com/troubleshooting-ssl.html).

We cannot suggest much more with the amount of information provided. At the very least, please provide the effective configuration and logs. We do not guess in this community.

**Comment by michaelklishin** [comment]
Also, I'm pretty sure combining `mqtt.allow_anonymous = true` and x.509-certificate based authentication will not work.

`mqtt.allow_anonymous` instructs the plugin to use its default credentials. Avoid anonymous "authentication" in MQTT at all costs either way.

**Comment by michaelklishin** [comment]
Oh, and the most obvious thing to verify is what your client certificate's Subject Alternative Name and the CN values are, and that there there is a user with that username in RabbitMQ's internal database, since that's the authN backend (not to be confused with mechanisms) used by your node (according to the provided configuration file).

**Comment by lukebakken** [comment]
@Aleksandar1932 You're in luck, I have a **complete** example here:

https://github.com/lukebakken/mqtt-client-cert-example

See the README and `setup.sh` files for how to configure everything correctly. Let me know if you have questions.

**Comment by rukechen** [comment]
I got one familiar case. I just used RabbitMQ and didn't enable mqtt plugin.
The following is my configuration.
`management.listener.port = 15672
management.listener.ssl = true
auth_backends.1 = cache
#
auth_cache.cached_backend = http
auth_cache.cache_ttl = 3600000


auth_http.http_method   = get
auth_http.user_path     = http://localhost:3001/auth/user
auth_http.vhost_path    = http://localhost:3001/auth/vhost
auth_http.resource_path = http://localhost:3001/auth/resource
auth_http.topic_path    = http://localhost:3001/auth/topic
auth_backends.2 = internal
auth_mechanisms.1 = EXTERNAL
auth_mechanisms.2 = PLAIN
auth_mechanisms.3 = AMQPLAIN
ssl_cert_login_from = common_name
listeners.ssl.default = 5671
ssl_options.cacertfile = /etc/rabbitmq/ssl/ca_certificate.pem
ssl_options.certfile   = /etc/rabbitmq/ssl/server_certificate.pem
ssl_options.keyfile    = /etc/rabbitmq/ssl/server_key.pem
ssl_options.verify     = verify_peer
ssl_options.fail_if_no_peer_cert = true

management.listener.ssl_opts.cacertfile = /etc/rabbitmq/ssl/ca_certificate.pem
management.listener.ssl_opts.certfile = /etc/rabbitmq/ssl/server_certificate.pem
management.listener.ssl_opts.keyfile = /etc/rabbitmq/ssl/server_key.pem
management.listener.ssl_opts.verify = verify_peer
`
According to this page, https://www.rabbitmq.com/troubleshooting-ssl.html#openssl-tools
I think I should see
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/f0ee1604-b189-42d3-b3de-02821f7b7ba7)

but actually I only get this
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/1d02147e-c283-4261-b96a-0589b950b803)

Does RabbitMQ without mqtt plugin support x509 authentication? or is there detail standard operation process ?

Thanks

**Reply by lukebakken** [reply]
> Does RabbitMQ without mqtt plugin support x509 authentication?

Yes, of course it does.

* Have you enabled the `rabbitmq_auth_mechanism_ssl` plugin?
* Did you correctly implement X509 certificate authentication in your application? Without seeing your code, there is no assistance I can give with this. `handshake_timeout` usually means there is a bug in your application.
* Did you add a password-less user whose name is equal to the `CN=` (common name) value in your client certificate?

**Reply by rukechen** [reply]
My common name is CN=myclient
openssl x509 -in /certs/client/cert.pem -text
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/29909d86-11b3-4a9e-85aa-664028ecc8fc)

So I create user and set permission by the following command.
`rabbitmqctl add_user CN=myclient`
`rabbitmqctl set_user_tags CN=myclient administrator`
`rabbitmqctl set_permissions -p / CN=myclient ".*" ".*" ".*"`

I don't input the password field, make it as blank. Does that what you means password-less?

I also add auth_backends.3 = amqp in rabbitmq.conf.

After rabbitmq server is running, I use the following command to test.
curl -vvv --cert /certs/client/cert.pem  --key /certs/client/key.pem --cacert "/certs/ca/cacert.pem" 'https://CN=myclient@127.0.0.1:15672/api/overview'

The result is as following:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/dfd50c05-4fd2-4636-9f69-c72ba03219c8)

By the way, this is working https://github.com/lukebakken/rabbitmq-users-CzgiqOUso2k. But it seems not password-less.

**Reply by michaelklishin** [reply]
If you use x.509 certificates for authentication, the password mechanism is not used and the password value will be ignored by RabbitMQ nodes.

The password field does not disappear from the protocol, so yes, it will be set to something even in that case. You can set it to a randomly generated string if that would alleviate your concerns.

**Reply by rukechen** [reply]
Hi @michaelklishin ,
Do you mean rabbitmqctl add_user CN=myclient with password or call curl -vvv --cert /certs/client/cert.pem  --key /certs/client/key.pem --cacert "/certs/ca/cacert.pem" 'https://CN=myclient@127.0.0.1:15672/api/overview' with password?

I try call curl -vvv --cert /certs/client/cert.pem  --key /certs/client/key.pem --cacert "/certs/ca/cacert.pem" 'https://CN=myclient@127.0.0.1:15672/api/overview' with password and still get {"error":"not_authorized","reason":"Login failed"}.

![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/52e3e33e-1894-42e2-99f7-1d7db02f1027)

![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/5aa1b98d-91f2-40e9-bac9-1fbb765b08d5)

**Reply by michaelklishin** [reply]
I was referring to client connections. You are trying to use x.509 authentication with the HTTP API.

It does not support this option. [Protocol plugins do](https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_auth_mechanism_ssl), as described in the `rabbitmq_auth_mechanism_ssl` backend.

HTTP API clients must use either a username/password pair or JWT tokens.

**Reply by rukechen** [reply]
Got it. 
I will try to use client connection, not HTTP API.
Thanks

**Reply by rukechen** [reply]
I just port 5671 to send message, but still get authentication failed.
I try to use CN=myclient only or CN=myclient:random_pwd，the result is as following. 
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/425325e5-f7bd-4c1a-97c2-3b8c89673969)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/18527508/b3823cd5-c1e8-4f5c-879a-6b2d39ce5b7c)

**Reply by michaelklishin** [reply]
According to the debug log, `rabbit_auth_backend_amqp:user_login_authentication/2` is an undefined function (the error says "undef"). 

The function [clearly does exist](https://github.com/rabbitmq/rabbitmq-auth-backend-amqp/blob/main/src/rabbit_auth_backend_amqp.erl#L17), so perhaps the plugin is not enabled.

This must be the first time in a decade when I see someone try to combine `rabbit_auth_mechanism_ssl` and `rabbit_auth_backend_amqp`. The latter is significantly less popular than the three other [common authN/authZ options](https://rabbitmq.com/access-control.html#backends).
